src/main/java/io/deepstream/utils/AckTimeoutRegistry.java

```diff
@@ -0,0 +1,167 @@
+     * so is worth flagging when used ).
+     */
+    public void add( String name ) {
+        this.add( name, null );
+    }
+
+    /**
+     * Checks to see if an ack timer already exists in the register for
+     * the given name and action. If it does, it clears it, then starts a new one.
+     *
+     * @param name The name to be added to the register
+     * @param action The action to be added to the register
+     */
+    public void add( String name, Actions action ) {
+        String uniqueName = this.getUniqueName( action, name );
+        AckTimeoutTask task = this.register.get( uniqueName );
+        if( task != null ) {
+            clear( uniqueName );
+        }
+        addToRegister( uniqueName );
+    }
+
+    @Override
+    public void connectionStateChanged(ConnectionState connectionState) {
+        if( connectionState == ConnectionState.OPEN ) {
+            scheduleAcks();
+        }
+        this.state = connectionState;
+    }
+
+    /**
+     * Clears the ack timeout for a message.
+     *
+     * @param uniqueName The name of the message ( and possible action ) to remove the timeout for
+     */
+    private boolean clear( String uniqueName ) {
+        AckTimeoutTask task = register.get( uniqueName );
+        if( task != null ) {
+            task.cancel();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Adds the uniqueName to the register. Only schedules the timer if
+     * the connection state is OPEN, otherwise it adds to the queue of waiting acks.
+     *
+     * @param uniqueName The name to be added to the register
+     */
+    private void addToRegister( String uniqueName ) {
+        AckTimeoutTask task = new AckTimeoutTask( uniqueName, this );
+        register.put( uniqueName, task );
+
+        if( this.state == ConnectionState.OPEN ) {
+            timer.schedule( task, this.timeoutDuration );
+        } else {
+            this.ackTimers.add( task );
+        }
+    }
+
+    private void onTimeout( String name ) {
+        this.register.remove( name );
+        String msg = "No ACK message received in time for " + name;
+        this.client.onError( this.topic, Event.ACK_TIMEOUT, msg );
+    }
+
+    private void scheduleAcks() {
+        AckTimeoutTask task = null;
+        while( this.ackTimers.peek() != null ) {
+            try {
+                task = this.ackTimers.take();
+            } catch (InterruptedException e) {}
+
+            if( task != null ) {
+                this.timer.schedule( task, this.timeoutDuration );
+            }
+            task = null;
+        }
+    }
+
+    private String getUniqueName( Actions action, String name ) {
+        if( action != null ) {
+            return action.toString() + name;
+        } else {
+            return name;
+        }
+    }
+
+    private class AckTimeoutTask extends TimerTask {

```

Comment:
Moved as an inner class since it will never be access by outside world ( or tests )